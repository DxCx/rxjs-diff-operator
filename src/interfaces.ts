export type IObservableDiffType = "init" | "next" | "error" | "complete";

export interface IObservableDiff {
    type: IObservableDiffType;
    payload?: any;
    isObject?: boolean;
}
