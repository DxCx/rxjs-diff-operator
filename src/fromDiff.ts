import { Observable } from 'rxjs/Observable';
import { Operator } from 'rxjs/Operator';
import { Subscriber } from 'rxjs/Subscriber';
import { IObservableDiff } from './interfaces';
import { applyChange } from 'deep-diff';
import * as deepFreeze from 'deep-freeze';
import * as clone from 'clone';

export interface FromDiffSignature<T> {
  (): Observable<IObservableDiff>;
}

export function fromDiff<T extends IObservableDiff, R>(): Observable<R> {
  return this.lift(new FromDiffOperator<T>());
}

export class FromDiffOperator<T extends IObservableDiff> implements Operator<T, T> {
  public call(subscriber: Subscriber<T>, source: any): any {
    return source._subscribe(new FromDiffSubscriber(subscriber));
  }
}

class FromDiffSubscriber<T extends IObservableDiff> extends Subscriber<T> {
  private count: number = 0;
  private lastValue: any;

  constructor(destination: Subscriber<T>) {
    super(destination);
  }

  private _updateValue(newValue: any, isObject: boolean) {
      this.lastValue = clone(newValue);
      deepFreeze(this.lastValue);
      this.destination.next(isObject ? this.lastValue : this.lastValue.p);
  }

  private _process_diff({type, payload, isObject}: IObservableDiff) {
      if ( 0 === this.count ) {
          if ( type !== "init") {
              this.destination.error(new Error('Init message was not emitted.'));
              return;
          }
          this._updateValue(payload, isObject);
          return;
      }

      switch ( type ) {
          case "init":
              this.destination.error(new Error('Init message emitted while in sequance'));
              break;
          case "update":
              let copyValue = clone(this.lastValue);
              payload.forEach((changeset: any) => {
                  applyChange(copyValue, true, changeset);
              });
              this._updateValue(copyValue, isObject);
              break;
          case "error":
              this.destination.error(new Error(payload));
              break;
          case "complete":
              this.destination.complete();
              break;
      }
  }

  protected _next(value: IObservableDiff) {
      this._process_diff(value);
      this.count ++;
  }
}

Observable.prototype.fromDiff = fromDiff;

declare module 'rxjs/Observable' {
  interface Observable<T> {
    fromDiff: FromDiffSignature<T>;
  }
}
