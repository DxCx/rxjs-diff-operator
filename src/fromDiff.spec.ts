'use strict';

import 'jest';
require('babel-core/register');
require('babel-polyfill');

import { Observable } from 'rxjs';
import { IObservableDiff } from './interfaces';
import './fromDiff';

describe('fromDiff operator', () => {
    it('Should be pass sanity', () => {
        expect(typeof (<any>Observable.prototype).fromDiff).toBe('function');
    });

    it('emits basic changes', () => {
        let obs: Observable<IObservableDiff> = Observable.of({
            "isObject": false,
            "payload": {
                "p": 1
            },
            "type": "init"
        }, {
            "isObject": false,
            "payload": [
                {
                    "kind": "E",
                    "lhs": 1,
                    "path": [
                        "p"
                    ],
                    "rhs": 2
                }
            ],
            "type": "update"
        }, {
            "isObject": false,
            "payload": [
                {
                    "kind": "E",
                    "lhs": 2,
                    "path": [
                        "p"
                    ],
                    "rhs": 3
                }
            ],
            "type": "update"
        }, {
            "isObject": false,
            "payload": [
                {
                    "kind": "E",
                    "lhs": 3,
                    "path": [
                        "p"
                    ],
                    "rhs": 4
                }
            ],
            "type": "update"
        }, {
            "isObject": false,
            "payload": [
                {
                    "kind": "E",
                    "lhs": 4,
                    "path": [
                        "p"
                    ],
                    "rhs": 5
                }
            ],
            "type": "update"
        }, {
            "type": "complete"
        });

        return obs.fromDiff().bufferCount(6).toPromise().then(msgs => {
            expect(typeof msgs).toBe('object');
            expect(Array.isArray(msgs)).toBe(true);
            expect(msgs).toMatchSnapshot();
        });
    });
});
