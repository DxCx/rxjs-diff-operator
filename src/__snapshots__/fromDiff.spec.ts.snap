exports[`fromDiff operator emits basic changes 1`] = `
Array [
  1,
  2,
  3,
  4,
  5
]
`;
